### PR TITLE
Add REGEXP_REPLACE support (including unit test)

### LINKED ISSUE #47: Support regex_replace
[Torsten Landsiedel](https://twitter.com/zodiac1978) said:

> I would like to know if regexp_replace would be available in Playground
> Fixing the native search is my use:
> https://wordpress.org/plugins/ignore-block-name-in-search/
> But Regex_Replace is only available since MySQL 8.0.4+ and MariaDB 10.0.5+ and my check prevents the activation in Playground.

Let's support regex_replace

### PR DESCRIPTION
**Pull request by Zodiac1978**
Fixes #47

### CODE REVIEW

File: tests/WP_SQLite_Translator_Tests.php [hunk 1]
```diff
@@ -118,6 +118,23 @@ public static function regexpOperators() {
...
+			"INSERT INTO _options (option_name, option_value) VALUES ('test-remove', '2');"
+		);
+
+		$this->assertQuery( "SELECT * FROM _options WHERE REGEXP_REPLACE(option_name, '(-ignore|-remove)', '') = 'test'" );
+		$this->assertCount( 2, $this->engine->get_query_results() );
+
+		$result1 = $this->engine->query( "SELECT option_value FROM _options WHERE option_name='test-ignore';" );
+		$result2 = $this->engine->query( "SELECT option_value FROM _options WHERE option_name='test-remove';" );
+		$this->assertEquals( '1', $result1[0]->option_value );
+		$this->assertEquals( '2', $result2[0]->option_value );
```
Comment (Zodiac1978) [inline tests/WP_SQLite_Translator_Tests.php:135]:
```suggestion
```

**Comment by Zodiac1978** [comment]
Happy to get your review @bgrgicak ! :)

**Comment by Zodiac1978** [comment]
@adamziel I have now additionally added checks for if any of the required parameters are null.

Copying the result from MySQL:
> If expr, pat, or repl is NULL, the return value is NULL. 
https://dev.mysql.com/doc/refman/8.4/en/regexp.html#function_regexp-replace

File: tests/WP_SQLite_Translator_Tests.php [hunk 2]
```diff
@@ -118,6 +118,30 @@ public static function regexpOperators() {
...
+			"INSERT INTO _options (option_name, option_value) VALUES ('test-ignore', '1');"
+		);
+		$this->assertQuery(
+			"INSERT INTO _options (option_name, option_value) VALUES ('test-remove', '2');"
+		);
+
+		$this->assertQuery( "SELECT * FROM _options WHERE REGEXP_REPLACE(option_name, '(-ignore|-remove)', '') = 'test'" );
+		$this->assertCount( 2, $this->engine->get_query_results() );
+
+		$this->assertQuery( "SELECT REGEXP_REPLACE( null, 'a', 'x') as result" );
```
Comment (bgrgicak) [inline tests/WP_SQLite_Translator_Tests.php:132]: Let's add some comments to explain these edge cases

File: tests/WP_SQLite_Translator_Tests.php [hunk 3]
```diff
@@ -118,6 +118,30 @@ public static function regexpOperators() {
...
+		$this->assertEquals( null, $results[0]->result );
+
+		$this->assertQuery( "SELECT REGEXP_REPLACE( 'abc', null, 'x') as result" );
+		$results = $this->engine->get_query_results();
+		$this->assertEquals( null, $results[0]->result );
+
+		$this->assertQuery( "SELECT REGEXP_REPLACE( 'abc', 'a', null) as result" );
+		$results = $this->engine->get_query_results();
+		$this->assertEquals( null, $results[0]->result );
+	}
```
Comment (bgrgicak) [inline tests/WP_SQLite_Translator_Tests.php:143]: Can we add a test for empty string patterns?

**Review by bgrgicak — APPROVED** [review]
Thank you!

**Review by adamziel — APPROVED** [review]
Lgtm!

**Comment by adamziel** [comment]
Actually, let’s also test for empty string pattern, not just a null one

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php [hunk 4]
```diff
@@ -493,6 +494,54 @@ public function regexp( $pattern, $field ) {
...
+
+		/* Return null if one of the required parameter is null */
+		if ( is_null( $field ) || is_null( $pattern ) || is_null( $replacement ) ) {
+			return null;
+		}
+
+		/* Return null if the pattern is empty - this changes MySQL/MariaDB behavior! */
+		if ( empty( $pattern ) ) {
+			return null;
+		}
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:530]: Oh! Wha is the MySQL/MariaDB behavior? to just use an empty pattern? If so, let's do that.

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php [hunk 5]
```diff
@@ -493,6 +494,54 @@ public function regexp( $pattern, $field ) {
...
+		 * byte codes.
+		 *
+		 * The REGEXP function can't be easily made to accept two
+		 * parameters, so we'll have to use a hack to get around this.
+		 *
+		 * If the first character of the pattern is a null byte, we'll
+		 * remove it and make the comparison case-sensitive. This should
+		 * be reasonably safe since PHP does not allow null bytes in
+		 * regular expressions anyway.
+		 */
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:520]: Does regexp_replace support the `BINARY` keyword? I couldn't find anything. If it doesn't, let's remove this comment and the special casing for the null byte.

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php [hunk 6]
```diff
@@ -493,6 +494,54 @@ public function regexp( $pattern, $field ) {
...
+	/**
+	 * Method to emulate MySQL REGEXP_REPLACE() function.
+	 *
+	 * @param string|array $pattern     Regular expression to search for (or array of strings).
+	 * @param string|array $replacement The string or an array with strings to replace.
+	 * @param string|array $field       Haystack.
+	 *
+	 * @return Array if the field parameter is an array, or a string otherwise.
+	 */
+	public function regexp_replace( $field, $pattern, $replacement ) {
```
Comment (adamziel) [inline wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:506]: I just checked and in MySQL REGEXP_REPLACE() also [takes three optional parameters](https://dev.mysql.com/doc/refman/8.4/en/regexp.html#function_regexp-replace):

> pos: The position in expr at which to start the search. If omitted, the default is 1.
> occurrence: Which occurrence of a match to replace. If omitted, the default is 0 (which means “replace all occurrences”).
> match_type: A string that specifies how to perform matching. The meaning is as described for [REGEXP_LIKE()](https://dev.mysql.com/doc/refman/8.4/en/regexp.html#function_regexp-like).

Would you be up for adding a support for those? If not, that's fine, but let's at add them to the function signature and fail with an informative warning (`REGEXP_REPLACE() don't support non-default values for $pos (fourth argument), .... if you need it then you can contribute here: ....`) if they're set to non-default values.

**Comment by adamziel** [comment]
I just had a moment to review this in a more focused setting and I left a few more notes. It is pretty close!

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php (hunk 4 above)
Comment (Zodiac1978) [reply wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:530]: It is throwing an error (because without pattern it can't work at all):
> Error in the SQL query (3685): Illegal argument to a regular expression. 

@bgrgicak was suggesting to return null instead to avoid breaking things, I assume.

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php (hunk 6 above)
Comment (Zodiac1978) [reply wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:506]: > Would you be up for adding a support for those?

I will try to do that!

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php (hunk 5 above)
Comment (Zodiac1978) [reply wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:520]: Isn't the binary-mode something global?
See: https://mariadb.com/docs/server/ref/mdb/cli/mariadb/binary-mode/

File: wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php (hunk 4 above)
Comment (bgrgicak) [reply wp-includes/sqlite/class-wp-sqlite-pdo-user-defined-functions.php:530]: Null seemed like a better response from an error. @Zodiac1978 let's update the comment to make it clear what _this changes MySQL/MariaDB behavior_ means.

**Comment by bgrgicak** [comment]
@JanJakes is this PR still relevant?

**Comment by JanJakes** [comment]
@bgrgicak I think it is; we only need to update the tests as well for the new driver. I can take a look at this.

**Comment by JanJakes** [comment]
Closing this in favor of https://github.com/WordPress/sqlite-database-integration/pull/368.
